src/protocols/rdp/download.c

```diff
@@ -0,0 +1,240 @@
+    int file_id = guac_rdp_fs_open(fs, name, GENERIC_READ, 0, FILE_OPEN, 0);
+    if (file_id < 0) {
+        guac_user_log(user, GUAC_LOG_INFO, "Unable to read file \"%s\"",
+                name);
+        return 0;
+    }
+
+    /* Get opened file */
+    guac_rdp_fs_file* file = guac_rdp_fs_get_file(fs, file_id);
+    if (file == NULL) {
+        guac_client_log(fs->client, GUAC_LOG_DEBUG,
+                "%s: Successful open produced bad file_id: %i",
+                __func__, file_id);
+        return 0;
+    }
+
+    /* If directory, send contents of directory */
+    if (file->attributes & FILE_ATTRIBUTE_DIRECTORY) {
+
+        /* Create stream data */
+        guac_rdp_ls_status* ls_status = malloc(sizeof(guac_rdp_ls_status));
+        ls_status->fs = fs;
+        ls_status->file_id = file_id;
+        guac_strlcpy(ls_status->directory_name, name,
+                sizeof(ls_status->directory_name));
+
+        /* Allocate stream for body */
+        guac_stream* stream = guac_user_alloc_stream(user);
+        stream->ack_handler = guac_rdp_ls_ack_handler;
+        stream->data = ls_status;
+
+        /* Init JSON object state */
+        guac_common_json_begin_object(user, stream,
+                &ls_status->json_state);
+
+        /* Associate new stream with get request */
+        guac_protocol_send_body(user->socket, object, stream,
+                GUAC_USER_STREAM_INDEX_MIMETYPE, name);
+
+    }
+
+    /* Otherwise, send file contents */
+    else {
+
+        /* Create stream data */
+        guac_rdp_download_status* download_status = malloc(sizeof(guac_rdp_download_status));
+        download_status->file_id = file_id;
+        download_status->offset = 0;
+
+        /* Allocate stream for body */
+        guac_stream* stream = guac_user_alloc_stream(user);
+        stream->data = download_status;
+        stream->ack_handler = guac_rdp_download_ack_handler;
+
+        /* Associate new stream with get request */
+        guac_protocol_send_body(user->socket, object, stream,
+                "application/octet-stream", name);
+
+    }
+
+    guac_socket_flush(user->socket);
+    return 0;
+}
+
+void* guac_rdp_download_to_user(guac_user* user, void* data) {
+
+    /* Do not bother attempting the download if the user has left */
+    if (user == NULL)
+        return NULL;
+
+    guac_client* client = user->client;
+    guac_rdp_client* rdp_client = (guac_rdp_client*) client->data;
+    guac_rdp_fs* filesystem = rdp_client->filesystem;
+
+    /* Ignore download if filesystem has been unloaded */
+    if (filesystem == NULL)
+        return NULL;
+
+    /* Attempt to open requested file */
+    char* path = (char*) data;
+    int file_id = guac_rdp_fs_open(filesystem, path,
+            FILE_READ_DATA, 0, FILE_OPEN, 0);
+
+    /* If file opened successfully, start stream */
+    if (file_id >= 0) {
+
+        guac_rdp_download_status* download_status;
+        const char* basename;
+
+        int i;
```

Comment:
Why not declare these variables down where they're actually used?

Reply:
It was the coding style at the time that this function was originally written. The body of this function was actually moved from what used to be `rdpdr_service.c` and was originally called `guac_rdpdr_download_to_owner`:

https://github.com/apache/guacamole-server/blob/f5b5ac718396b24352290d2fac976260f4b3829e/src/protocols/rdp/guac_rdpdr/rdpdr_service.c#L237-L303

I can clean this up a bit now in the locations I can find such code.

Reply:
OK - I've cleaned this up, moved the basename logic to its own function within `fs.h`, and added a unit test for said function.